lib/puppetserver/ca/certificate_authority.rb

```diff
@@ -0,0 +1,210 @@
+require 'puppetserver/ca/utils/http_client'
+
+require 'json'
+
+module Puppetserver
+  module Ca
+    class CertificateAuthority
+
+      include Puppetserver::Ca::Utils
+
+      REVOKE_BODY = JSON.dump({ desired_state: 'revoked' })
+      SIGN_BODY =   JSON.dump({ desired_state: 'signed' })
+
+      def initialize(logger, settings)
+        @logger = logger
+        @client = HttpClient.new(settings)
+        @ca_server = settings[:ca_server]
+        @ca_port = settings[:ca_port]
+      end
+
+      # Returns a URI-like wrapper around CA specific urls
+      def make_ca_url(resource_type = nil, certname = nil)
+        HttpClient::URL.new('https', @ca_server, @ca_port, 'puppet-ca', 'v1', resource_type, certname)
+      end
+
+      def sign_certs(certnames)
+        put('certificate_status', certnames, SIGN_BODY, :sign)
+      end
+
+      def revoke_certs(certnames)
+        put('certificate_status', certnames, REVOKE_BODY, :revoke)
+      end
+
+      # Should this csr be in JSON? instead of string?
+      def submit_certificate_request(certname, csr)
+        put('certificate_request', [certname], csr, :submit)
+      end
+
+      # Make an HTTP PUT request to CA
+      # @param resource_type [String] the resource type of url
+      # @param certnames [Array] array of certnames
+      # @param body [JSON/String] body of the put request
+      # @param type [Symbol] type of error processing to perform on result
+      # @return [Boolean] whether all requests were successful
+      def put(resource_type, certnames, body, type)
+        url = make_ca_url(resource_type)
+        results = @client.with_connection(url) do |connection|
+          certnames.map do |certname|
+            url.resource_name = certname
+            result = connection.put(body, url)
+            process_results(type, certname, result)
+          end
+        end
+
+        results.all?
+      end
+
+      # logs the action and returns true/false for success
+      def process_results(type, certname, result)
+        case type
+        when :sign
+          case result.code
+          when '204'
+            @logger.inform "Successfully signed certificate request for #{certname}"
+            return true
+          else
+            @logger.err 'Error:'
+            @logger.err "    When signing request submitted for #{certname}:"
```

Comment:
It seems like almost all of these are variations on:
success:
```
@logger.inform "Was successful in #{verb} #{object} for #{certname}"
```
error:
```
@logger.err 'Error:'
@logger.err "    When #{verb} #{object} for #{certname} received:"
# ...
```

Would it make it easier to read to refactor that out? Or would do you think it'd be a bit too obscure?

Reply:
If we do it like this, it becomes impossible to translate. Since we might still want that for this gem one day, I think we should leave it more explicit.

Reply:
See https://confluence.puppetlabs.com/display/enterprise/Translation-friendly+strings for guidelines on writing translatable strings.

Reply:
Good catch, and thanks for the link!

I imagine that making a giant table of strings and turning this into a bunch of selectors would make this impenetrable?